### PR TITLE
[Merged by Bors] - Fix client authentication and split tls directories

### PR DESCRIPTION
**Pull request by maltesander**
# Description

- Switched init container to tools image
- Fixed client authentication. Now only the provided secretClass is eligible. Split up tls (client/quorum) dirs into separate directories and create key/truststores in different directory

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/zk_controller.rs [hunk 1]
```diff
@@ -499,15 +499,19 @@ fn build_server_rolegroup_statefulset(
     server_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
     client_authentication_class: Option<&AuthenticationClass>,
 ) -> Result<StatefulSet> {
+    let mut cb_prepare = ContainerBuilder::new("prepare");
+    let mut cb_zookeeper = ContainerBuilder::new(APP_NAME);
+    let mut pod_builder = PodBuilder::new();
```
Comment (teozkr) [inline rust/operator-binary/src/zk_controller.rs:504]: Moving these up like this (away from their respective use sites) implies that they are reusable, but the `mut` gives away that they aren't.

File: rust/operator-binary/src/zk_controller.rs [hunk 2]
```diff
@@ -715,14 +672,96 @@ fn build_server_rolegroup_statefulset(
...
+            }
+            _ => {
+                return Err(Error::AuthenticationMethodNotSupported {
+                    authentication_class: ObjectRef::from_obj(auth_class),
+                    supported: vec!["tls".to_string()],
+                    method: auth_class.spec.provider.to_string(),
+                })
+            }
+        }
+    } else if let Some(client_tls) = zk.client_tls_secret_class() {
```
Comment (teozkr) [inline rust/operator-binary/src/zk_controller.rs:716]: Sounds like it should be an error to set TLS simultaneously via a SecretClass and AuthenticationClass?

Comment (teozkr) [reply rust/operator-binary/src/zk_controller.rs:716]: This logic is also very similar.. would it make sense to translate direct TLS config into a synthetic `AuthenticationClass` before we get this far (or the reverse)?

File: rust/crd/src/lib.rs [hunk 3]
```diff
@@ -412,25 +416,25 @@ impl ZookeeperCluster {
     }
 
     /// Checks if we should use TLS to encrypt client connections.
-    pub fn is_client_secure(&self) -> bool {
-        self.client_tls_secret_class().is_some()
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs:416]: Yes it does. The `spec.config.tls` must be set explicitly to null (default is "tls"). I dont like that behavior to be honest (and therefore not like in trino where no tls auth is implemented i would leave it out here)

File: rust/crd/src/lib.rs [hunk 4]
```diff
@@ -412,25 +416,25 @@ impl ZookeeperCluster {
...
+    pub fn tls_enabled(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs]: done

File: rust/operator-binary/src/zk_controller.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator-binary/src/zk_controller.rs]: done

File: rust/operator-binary/src/zk_controller.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator-binary/src/zk_controller.rs:704]: The secret_class is now extracted beforehand:
```
    let tls_secret_class = if let Some(auth_class) = client_authentication_class {
        match &auth_class.spec.provider {
            AuthenticationClassProvider::Tls(TlsAuthenticationProvider {
                client_cert_secret_class: Some(secret_class),
            }) => Some(secret_class),
            _ => {
                return Err(Error::AuthenticationMethodNotSupported {
                    authentication_class: ObjectRef::from_obj(auth_class),
                    supported: vec!["tls".to_string()],
                    method: auth_class.spec.provider.to_string(),
                })
            }
        }
    } else if let Some(client_tls) = zk.client_tls_secret_class() {
        Some(&client_tls.secret_class)
    } else {
        None
    };
```

File: rust/crd/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (teozkr) [reply rust/crd/src/lib.rs:416]: Huh?

[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs:416]: About fixing / failing invalid configs. I added to the docs that authentication takes precedence over just tls.

File: rust/operator-binary/src/command.rs [hunk 5]
```diff
@@ -21,43 +20,50 @@ pub fn create_init_container_command_args(zk: &ZookeeperCluster) -> String {
...
+        args.extend(create_key_and_trust_store_cmd(
+            CLIENT_TLS_MOUNT_DIR,
+            CLIENT_TLS_DIR,
+            "client-tls",
+        ));
+
+        // only copy system truststore if client tls enabled and no client auth specified
+        if zk.client_tls_secret_class().is_some() && zk.client_tls_authentication_class().is_none()
+        {
+            args.push(format!("keytool -importkeystore -srckeystore {SYSTEM_TRUST_STORE_DIR} -srcstoretype jks -srcstorepass changeit -destkeystore {CLIENT_TLS_DIR}/truststore.p12 -deststoretype pkcs12 -deststorepass ${STORE_PASSWORD_ENV} -noprompt"));
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/command.rs]: yeah removed it again. Confused the client tls stuff.

**Review by teozkr — APPROVED** [review]
Untested but LGTM, ok to merge if tests pass

**Pull request closed without merging**